### ISSUE TITLE
 Do 'ulimit -n 5328' in the console to overcome the limit

### ISSUE THREAD

**Issue by tseemann** [ablab/spades#120]
it seems Spades 3.12.0 needs to open too many files at once?

```
)   Failed to setup necessary limit for number of open files. The process might crash later on.
[spades]  * 0:00:13.079     8M / 564M  WARN    General                 (kmer_splitters.hpp        :  83)   Do 'ulimit -n 5328' in the console to overcome the limit
```
The hard limit on RHEL/Centos is 4096 files.

**Comment by asl** [comment]
Yes, each thread would open certain number of files (and could keep them open for some time), so the hard open file limit (and such low!) does not make much sense.

This part of the code was there for several years as of now and not 3.12 specific.

**Comment by tseemann** [comment]
Interesting, I had never encountered it before. I suspect because of the `OMP_THREAD_LIMIT` ignoring `--threads 72` and keeping threads down.  

1024 (soft) and 4096 (hard) open file limit is very common on most Unix.   Having it unlimited or very high is a potential security/stability risk, as one can create file/fork bombs and bring nodes down.

I agree it should be higher than 4096 (given that `/proc/sys/fs/file-max` is > 6 million), but i don't have root on my HPC :-(

Have you ever run Spades on 64+ cores?
What is your default ulimit -n ?

**Comment by asl** [comment]
I double checked the code and I believe these days we do not open that much # of files (though the check remains). So, for now it probably safe to ignore it. The code in question ought to be refactored some time soon

**Comment by tseemann** [comment]
Are you saying it is forcing me to change a root system setting that it doesn't actually need?

I can't ignore the error, because it prevents me from running Spades with lots of CPUs?

**Comment by asl** [comment]
Well, this is a very common misconfiguration of multi-core system. If you're having, say, 128 CPU cores and each thread would have, say, 10 open files, then we'd need to have at least 1.2k+ files open at the same time. So, the defaults should certainly be tuned on such systems during the system installation.

Again, I believe we fixed the case when the # of open files was very high, but the check remains just in case, because we have not inspected all such places, only the ones that certainly opened lots of files at the same time.

**Comment by tseemann** [comment]
When you say fixed, do you mean Spades 3.13.0 ?

**Comment by asl** [comment]
Probably around SPAdes 3.1 or sometime ~3-4 year ago.

**Comment by tseemann** [comment]
I am still confused. 

How many files per thread does Spades >= 3.12 need to open?

**Comment by asl** [comment]
C * # of threads. The C is not well defined yet.

**Comment by tseemann** [comment]
Can you put an upper and lower bound on C yet ? :)

**Comment by asl** [comment]
Not yet. Probably ~20-30

**Comment by tseemann** [comment]
Ok, so 72 threads x 30 = 2100 files. So `ulimit -n 5328` shouldn't be needed?

**Comment by asl** [comment]
Maybe. But it's not 100% sure.

**Comment by tseemann** [comment]
Ok thank you for answering.  

You might need to buy one of these for "testing" :-)
https://www.intel.com.au/content/www/au/en/products/processors/core/x-series/i9-7980xe.html
